### PR TITLE
server: fix memory trace's leak metrics

### LINKED ISSUE #15357: Memory Trace does not include alive or leak data for raft_router/apply_router
### What version of TiKV are you using?
master

### Steps to reproduce

We defined the raft_router_leak in tikv code, but that metrics' data is not reported to grafana. 

### What did you expect?
Screen Shot 2023-08-18 at 4 38 29 PM

### What did happened?
alive and leak data are reported together.

### PR DESCRIPTION
**Pull request by tonyxuqqi**
### What is changed and how it works?

Issue Number: Close #15357

What's Changed:

```commit-message
Correct the raft_router/apply_router's alive and leak metrics. 
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
- No code

Side effects

No

### Release note 

```release-note
Separate the apply_router, raft_router's metrics into alive and leak respectively. 
```

### CODE REVIEW

**Comment by tonyxuqqi** [comment]
vEz0hCBQRA

File: components/server/src/memory.rs [hunk 1]
```diff
@@ -13,15 +13,32 @@ pub struct MemoryTraceManager {
...
+                    let leak = sub_trace.sub_trace(Id::Name("leak"));
+                    MEM_TRACE_SUM_GAUGE
+                        .with_label_values(&[&format!(
+                            "{}-{}-alive",
+                            provider_name, sub_trace_name
+                        )])
+                        .set(alive.sum() as i64);
+
+                    MEM_TRACE_SUM_GAUGE
+                        .with_label_values(&[&format!("{}-{}-leak", provider_name, sub_trace_name)])
```
Comment (overvenus) [inline components/server/src/memory.rs:35]: `with_label_values` may acquire a lock here, is `flush` called frequently?

**Review by overvenus — APPROVED** [review]

File: components/server/src/memory.rs [hunk 2]
```diff
@@ -13,15 +13,32 @@ pub struct MemoryTraceManager {
...
             let provider_name = provider.name();
             let ids = provider.get_children_ids();
             for id in ids {
                 let sub_trace = provider.sub_trace(id);
                 let sub_trace_name = sub_trace.name();
-                MEM_TRACE_SUM_GAUGE
-                    .with_label_values(&[&format!("{}-{}", provider_name, sub_trace_name)])
-                    .set(sub_trace.sum() as i64)
+                if sub_trace_name == raft_router || sub_trace_name == apply_router {
+                    let alive = sub_trace.sub_trace(Id::Name("alive"));
```
Comment (Connor1996) [inline components/server/src/memory.rs:25]: Can we make it more generic, like using `get_children_ids` to record sub-traces recursively?

File: components/server/src/memory.rs [hunk 3]
```diff
@@ -13,15 +13,32 @@ pub struct MemoryTraceManager {
 
 impl MemoryTraceManager {
     pub fn flush(&mut self, _now: Instant) {
+        let raft_router = "raft_router".to_string();
```
Comment (LykxSassinator) [inline components/server/src/memory.rs:16]: Better to use `CONST` on these two strings.

File: components/server/src/memory.rs (hunk 1 above)
Comment (LykxSassinator) [reply components/server/src/memory.rs:35]: FYI, it's 1 second-level timer for flushing this metrics.
https://github.com/tikv/tikv/blob/3ae1fb4320737c71a1c9d3f8ee6a3b7a9af6f6ea/components/server/src/server.rs#L225
Maybe should be longer for this update?

**Review by LykxSassinator — APPROVED** [review]
Rest LGTM

File: components/server/src/memory.rs (hunk 2 above)
Comment (tonyxuqqi) [reply components/server/src/memory.rs]: Good point.

File: components/server/src/memory.rs [hunk 4]
```diff
@@ -19,9 +19,24 @@ impl MemoryTraceManager {
             for id in ids {
                 let sub_trace = provider.sub_trace(id);
                 let sub_trace_name = sub_trace.name();
-                MEM_TRACE_SUM_GAUGE
-                    .with_label_values(&[&format!("{}-{}", provider_name, sub_trace_name)])
-                    .set(sub_trace.sum() as i64)
+                let leaf_ids = sub_trace.get_children_ids();
```
Comment (Connor1996) [inline components/server/src/memory.rs:22]: How about
```rust
pub fn flush(&mut self, _now: Instant) {
        fn record = |trace: Arc<MemoryTrace>, names: &Vec<String>| {
            let name = trace.name();
            names.push(name);
            let subs = trace.get_children_ids();
            if subs.is_empty {
                MEM_TRACE_SUM_GAUGE
                    .with_label_values(&[&names.join("-")])
                    .set(trace.sum() as i64)
            } else {
                for sub in subs {
                    record(sub, names);
                }
            }
            names.pop();
        };
        let mut names = vec![];
        for provider in &self.providers {
            record(provider, &names);
        }
}
```

Comment (tonyxuqqi) [reply components/server/src/memory.rs:22]: In general, recursive function is not recommended in production code. We want to limit the max level of the metrics.

File: components/server/src/memory.rs (hunk 3 above)
Comment (tonyxuqqi) [reply components/server/src/memory.rs]: Good catch. They're all removed with a more general way.

File: components/server/src/memory.rs (hunk 1 above)
Comment (tonyxuqqi) [reply components/server/src/memory.rs]: I think it's ok. No reason to update it.

**Pull request merged**